File: lab3/src/main/java/lab3/functionality/Quest.java
package lab3.functionality;

public abstract class Quest {
    protected final Designer designer;
    protected final Player player;

    public Quest(Designer designer, Player player) {
        this.designer = designer;
        this.player = player;
    }

    public void start() {
        while (!player.isCompletedQuest()) {
            designer.interactionSession();
            checkPlayer();
        }
    }

    private void end(String reason) {
        designer.endMessage(reason);
        player.setCompletedQuest(true);
    }

    private void checkPlayer() {
        if (player.getHealth() == Health.DEAD) {
            end(String.format("%s умер!", player.getName()));
        } else if (player.isCompletedQuest()){
            end("Поздравляем с завершением квеста");
        }
    }
}

File: lab4/src/main/kotlin/lab4/functionality/Quest.kt
package lab4.functionality

class Quest(private val designer: Designer, protected val player: Player) {

    fun start() {
        while (!player.completedQuest) {
            designer.interactionSession()
            checkPlayer()
        }
    }

    private fun end(reason: String) {
        designer.endMessage(reason)
        player.completedQuest = true
    }

    private fun checkPlayer() {
        if (player.health === Health.DEAD) {
            end("$player.name умер!")
        } else if (player.completedQuest) {
            end("Поздравляем с завершением квеста")
        }
    }
}